### PR TITLE
Fixup: Initialising the filename variable

### PR DESCRIPTION
**Pull request by sathnaga**
Changing the directory inside test makes filename property set to None
results in resetting basedir and datadir to None, Hence initialize
the filename once.

Fixes: https://github.com/avocado-framework/avocado-misc-tests/issues/9

### CODE REVIEW

**Comment by sathnaga** [comment]
This patch creates a problem for running avocado-vt tests, needs a cleaner fix, ignore this.


2016-04-23 05:41:55,286 job              L0348 INFO |
2016-04-23 05:41:55,591 output           L0632 DEBUG| Process Process-1:
2016-04-23 05:41:55,591 output           L0632 DEBUG| Traceback (most recent call last):
2016-04-23 05:41:55,591 output           L0632 DEBUG|   File "/usr/lib64/python2.7/multiprocessing/process.py", line 258, in _bootstrap
2016-04-23 05:41:55,592 output           L0632 DEBUG|     self.run()
2016-04-23 05:41:55,592 output           L0632 DEBUG|   File "/usr/lib64/python2.7/multiprocessing/process.py", line 114, in run
2016-04-23 05:41:55,592 output           L0632 DEBUG|     self._target(*self._args, **self._kwargs)
2016-04-23 05:41:55,592 output           L0632 DEBUG|   File "/usr/lib/python2.7/site-packages/avocado-0.34.0-py2.7.egg/avocado/core/runner.py", line 232, in _run_test
2016-04-23 05:41:55,593 output           L0632 DEBUG|     instance = loader.load_test(test_factory)
2016-04-23 05:41:55,593 output           L0632 DEBUG|   File "/usr/lib/python2.7/site-packages/avocado-0.34.0-py2.7.egg/avocado/core/loader.py", line 263, in load_test
2016-04-23 05:41:55,593 output           L0632 DEBUG|     test_instance = test_class(**test_parameters)
2016-04-23 05:41:55,593 output           L0632 DEBUG|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-0.34.0-py2.7.egg/avocado_vt/test.py", line 113, in __init__
2016-04-23 05:41:55,602 output           L0632 DEBUG|     runner_queue=runner_queue)
2016-04-23 05:41:55,602 output           L0632 DEBUG|   File "/usr/lib/python2.7/site-packages/avocado-0.34.0-py2.7.egg/avocado/core/test.py", line 95, in __init__
2016-04-23 05:41:55,603 output           L0632 DEBUG|     self.filename = source
2016-04-23 05:41:55,603 output           L0632 DEBUG| AttributeError: can't set attribute

**Comment by ldoktor** [comment]
The filename used to be a property and it still is for `SimpleTest/ExternalTest`.

**Comment by ldoktor** [comment]
Hello Satheesh, it's the property, indeed. I went ahead and this is a very hackish way, which preserves the functionality:

```
diff --git a/avocado/core/test.py b/avocado/core/test.py
index ff115ca..c4f127b 100644
--- a/avocado/core/test.py
+++ b/avocado/core/test.py
@@ -87,15 +87,17 @@ class Test(unittest.TestCase):
             self.name = self.__class__.__name__
 
         # Initialize filename
-        possibly_compiled = inspect.getfile(self.__class__)
-        if possibly_compiled.endswith('.pyc') or possibly_compiled.endswith('.pyo'):
-            source = possibly_compiled[:-1]
-        else:
-            source = possibly_compiled
-        if os.path.exists(source):
-            self.filename = source
-        else:
-            self.filename = None
+        if not hasattr(self, 'filename'):
+            possibly_compiled = inspect.getfile(self.__class__)
+            if (possibly_compiled.endswith('.pyc') or
+                    possibly_compiled.endswith('.pyo')):
+                source = possibly_compiled[:-1]
+            else:
+                source = possibly_compiled
+            if os.path.exists(source):
+                self.filename = source
+            else:
+                self.filename = None
 
         self.tag = tag
         self.job = job
@@ -595,17 +597,12 @@ class SimpleTest(Test):
                                 r' \d\d:\d\d:\d\d WARN \|')
 
     def __init__(self, name, params=None, base_logdir=None, tag=None, job=None):
+        if not hasattr(self, 'filename'):
+            self.filename = os.path.abspath(name)
         super(SimpleTest, self).__init__(name=name, params=params,
                                          base_logdir=base_logdir, tag=tag, job=job)
         self._command = self.filename
 
-    @property
-    def filename(self):
-        """
-        Returns the name of the file (path) that holds the current test
-        """
-        return os.path.abspath(self.name)
-
     def _log_detailed_cmd_info(self, result):
         """
         Log detailed command information.
@@ -648,14 +645,12 @@ class ExternalRunnerTest(SimpleTest):
         self.assertIsNotNone(external_runner, "External runner test requires "
                              "external_runner parameter, got None instead.")
         self.external_runner = external_runner
+        if not hasattr(self, 'filename'):
+            self.filename = None
         super(ExternalRunnerTest, self).__init__(name, params, base_logdir,
                                                  tag, job)
         self._command = external_runner.runner + " " + name
 
-    @property
-    def filename(self):
-        return None
-
     def test(self):
         pre_cwd = os.getcwd()
         new_cwd = None
```

But I have to say I hate it. I remember @clebergnu introduced the `property` to protect and avoid such handling, so we should probably stick to it, but probably implement setter which sets in for the first time. Alternatively we can polish this version, it should also work.

**Comment by sathnaga** [comment]
@ldoktor : Yep, Agree with you, till we get the finer fix, we can go ahead with this hawkish way?, which should work for us, sending the updated patch and additionally the same has to be handled in avocado-vt too, ll send that too

File: avocado/core/test.py [hunk 1]
```diff
@@ -601,15 +598,10 @@ class SimpleTest(Test):
     def __init__(self, name, params=None, base_logdir=None, tag=None, job=None):
         super(SimpleTest, self).__init__(name=name, params=params,
                                          base_logdir=base_logdir, tag=tag, job=job)
+        if not hasattr(self, 'filename'):
+            self.filename = os.path.abspath(self.name)
```
Comment (ldoktor) [inline avocado/core/test.py]: You have to set this before calling the super class (basically just use the patch I provided, it passed the CI).

**Comment by ldoktor** [comment]
Hello @sathnaga, the reason CI is failing is because you call the superclass (which creates the filename attribute) and then ask if the filename attribute exists. It needs to be the other way around in order to work properly.

Anyway conceptually I agree it should fix the issue, unless @clebergnu disagrees (he had strong points about using the property)

**Comment by ldoktor** [comment]
Hello @sathnaga, would you please fix the travis issues so we can merge this? This worked for me, it's on top of ad83aaaf821ccea57a94d887dd4d2483047e021c, I'm not sure if it needs some cleanups and polishing. Also the avocado-vt might need to be updated, I haven't checked that:

```diff
commit 00892db7724cb35bdeaddaa54f98d3e1548a5c25
Author: Lukáš Doktor <ldoktor@redhat.com>
Date:   Fri Apr 22 17:42:14 2016 +0200

    My fixes
    
    Signed-off-by: Lukáš Doktor <ldoktor@redhat.com>

diff --git a/avocado/core/test.py b/avocado/core/test.py
index ff115ca..c4f127b 100644
--- a/avocado/core/test.py
+++ b/avocado/core/test.py
@@ -87,15 +87,17 @@ class Test(unittest.TestCase):
             self.name = self.__class__.__name__
 
         # Initialize filename
-        possibly_compiled = inspect.getfile(self.__class__)
-        if possibly_compiled.endswith('.pyc') or possibly_compiled.endswith('.pyo'):
-            source = possibly_compiled[:-1]
-        else:
-            source = possibly_compiled
-        if os.path.exists(source):
-            self.filename = source
-        else:
-            self.filename = None
+        if not hasattr(self, 'filename'):
+            possibly_compiled = inspect.getfile(self.__class__)
+            if (possibly_compiled.endswith('.pyc') or
+                    possibly_compiled.endswith('.pyo')):
+                source = possibly_compiled[:-1]
+            else:
+                source = possibly_compiled
+            if os.path.exists(source):
+                self.filename = source
+            else:
+                self.filename = None
 
         self.tag = tag
         self.job = job
@@ -595,17 +597,12 @@ class SimpleTest(Test):
                                 r' \d\d:\d\d:\d\d WARN \|')
 
     def __init__(self, name, params=None, base_logdir=None, tag=None, job=None):
+        if not hasattr(self, 'filename'):
+            self.filename = os.path.abspath(name)
         super(SimpleTest, self).__init__(name=name, params=params,
                                          base_logdir=base_logdir, tag=tag, job=job)
         self._command = self.filename
 
-    @property
-    def filename(self):
-        """
-        Returns the name of the file (path) that holds the current test
-        """
-        return os.path.abspath(self.name)
-
     def _log_detailed_cmd_info(self, result):
         """
         Log detailed command information.
@@ -648,14 +645,12 @@ class ExternalRunnerTest(SimpleTest):
         self.assertIsNotNone(external_runner, "External runner test requires "
                              "external_runner parameter, got None instead.")
         self.external_runner = external_runner
+        if not hasattr(self, 'filename'):
+            self.filename = None
         super(ExternalRunnerTest, self).__init__(name, params, base_logdir,
                                                  tag, job)
         self._command = external_runner.runner + " " + name
 
-    @property
-    def filename(self):
-        return None
-
     def test(self):
         pre_cwd = os.getcwd()
         new_cwd = None

```

**Comment by ldoktor** [comment]
Hello @sathnaga as the release is approaching and this is pretty serious fix I went ahead and squashed the changes and sent a new PR. I hope you don't mind, you're still listed as author.

**Comment by sathnaga** [comment]
@ldoktor : got busy into few other activities, actually, the changes were there already, After I rebased it got split into multiple commits. sorry for the inconvenience. Thanks :-)

**Comment by ldoktor** [comment]
No problem, actually @clebergnu found a cleaner solution, https://github.com/avocado-framework/avocado/pull/1217

**Comment by sathnaga** [comment]
@ldoktor Yep, have checked that, that is a sane, cleaner version :-)

**Pull request closed without merging**